File: python/honest-observe/src/honest_observe/browser.py
"""Browser instrumentation (section 8): the browser event envelope and the automatic browser events.

Browser-side events are beaconed to the same log as server events and joined to them by `request_id` —
one log, both sides. The pieces here are the pure data contracts: the envelope a browser event takes
(section 8.2) and the four events the DOM bootloader and domx emit automatically (section 8.4). The
beacon itself (`navigator.sendBeacon`) and the ingest endpoint that validates the session, stamps
`received_at`, and appends are boundary I/O — the browser event shape is what that endpoint validates
against and what tests assemble, exactly as `build_event` is the server envelope without the write.

The browser envelope differs from the server one: it carries `source: "browser"` and a `session_id`,
joins to server events through an optional `request_id`, and has no aggregate or sequence fields. The
impure values — a v4 `event_id` generated in the browser, a `timestamp` from the performance clock —
are passed in, so assembly stays pure.
"""

from honest_type import err, fault, ok

# A browser event cannot be valid without these (non-empty strings). `source` is the constant
# "browser"; `payload` may be empty; `request_id` is optional (absent before the first response).
_BROWSER_REQUIRED = ("event_id", "event_type", "event_version", "timestamp", "session_id")


def build_browser_event(event_type, event_version, timestamp, session_id, payload, event_id, request_id=None):
    """Assemble a validated browser event envelope (section 8.2). Pure. `source` is always "browser";
    `request_id` is attached only when supplied (it joins to the server events the browser event
    triggered). Returns ok(event), or err(fault 'invalid_event') naming any required field left empty."""
    event = {
        "event_id": event_id,
        "event_type": event_type,
        "event_version": event_version,
        "timestamp": timestamp,
        "source": "browser",
        "session_id": session_id,
        "payload": payload,
    }
    if request_id is not None:
        event["request_id"] = request_id
    missing = [name for name in _BROWSER_REQUIRED if not event[name]]
    if missing:
        return err(fault("invalid_event", f"Browser event is missing required field(s): {missing}", "client", {"missing": missing}))
    return ok(event)


def browser_classify(element, attribute, tokens, manifest, duration_ns, request_id=None):
    """The hf.browser.classify event payload (section 8.4): one attribute classification by the
    bootloader. `request_id` appears only within a request context. Pure."""
    payload = {"element": element, "attribute": attribute, "tokens": list(tokens), "manifest": manifest, "duration_ns": duration_ns}
    if request_id is not None:
        payload["request_id"] = request_id
    return {"event_type": "hf.browser.classify", "payload": payload}


def browser_request(method, url, trigger, target, manifest_keys, request_id):
    """The hf.browser.request event payload (section 8.4): one HTMX request from domx, carrying the
    request_id it sent so server events join to it. Pure."""
    return {"event_type": "hf.browser.request", "payload": {"method": method, "url": url, "trigger": trigger, "target": target, "manifest_keys": list(manifest_keys), "request_id": request_id}}


def browser_response(request_id, status, swap_target, duration_ms):
    """The hf.browser.response event payload (section 8.4): one HTMX response arriving, joined to its
    request and the server events by request_id. Pure."""
    return {"event_type": "hf.browser.response", "payload": {"request_id": request_id, "status": status, "swap_target": swap_target, "duration_ms": duration_ms}}


def dom_changed(changed_keys, from_values, to_values, request_id=None):
    """The hf.dom.changed event payload (section 8.4): a manifest state change seen by domx's mutation
    observer, with the previous and new values for the changed slots. `request_id` appears only when the
    change happens within a request context. Pure."""
    payload = {"changed_keys": list(changed_keys), "from": from_values, "to": to_values}
    if request_id is not None:
        payload["request_id"] = request_id
    return {"event_type": "hf.dom.changed", "payload": payload}

File: javascript/honest-dom/src/browser.js
// Browser observability (honest-DOM spec §5): domx is honest-observe's browser-side instrumentation.
// Every browser event is beaconed to /api/observe/ingest and joins the server log by request_id — one
// log, both sides. The event builders are pure and mirror honest-observe's browser.py contract field
// for field; emitBrowserEvent is the one boundary, taking the impure values (event_id, timestamp,
// session_id, request_id, mode) and sendBeacon through injected deps so assembly stays pure. request_id
// is never held in a module variable — that would be a second, hidden mutator of shared state. It lives
// in the DOM, written on each HTMX response and read fresh at emit time through deps.requestId (§5.2).

const EVENT_VERSION = "1.0";
const INGEST_ENDPOINT = "/api/observe/ingest";
// Payload fields that carry user-state values. Dropped outside development mode (§5.4): a browser event
// records which slots changed, never what they changed to.
const VALUE_BEARING = ["from", "to"];

// Assemble a browser event envelope (§5.1). Pure. source is always "browser"; request_id is attached
// only when supplied (null before the first response), since it joins to the server events triggered.
export function browserEvent(eventType, timestamp, sessionId, payload, eventId, requestId) {
  const event = {
    event_id: eventId,
    event_type: eventType,
    event_version: EVENT_VERSION,
    timestamp,
    source: "browser",
    session_id: sessionId,
    payload,
  };
  return requestId === null ? event : { ...event, request_id: requestId };
}

// The four automatic event payloads (§5.3), mirroring honest-observe's browser.py. Each is pure and
// returns { event_type, payload } — the shape emitBrowserEvent wraps in an envelope.

export function browserClassify(element, attribute, tokens, manifest, durationNs, requestId) {
  const payload = { element, attribute, tokens: [...tokens], manifest, duration_ns: durationNs };
  return requestId === null
    ? { event_type: "hf.browser.classify", payload }
    : { event_type: "hf.browser.classify", payload: { ...payload, request_id: requestId } };
}

export function browserRequest(method, url, trigger, target, manifestKeys, requestId) {
  return {
    event_type: "hf.browser.request",
    payload: { method, url, trigger, target, manifest_keys: [...manifestKeys], request_id: requestId },
  };
}

export function browserResponse(requestId, status, swapTarget, durationMs) {
  return {
    event_type: "hf.browser.response",
    payload: { request_id: requestId, status, swap_target: swapTarget, duration_ms: durationMs },
  };
}

export function domChanged(changedKeys, fromValues, toValues, requestId) {
  const payload = { changed_keys: [...changedKeys], from: fromValues, to: toValues };
  return requestId === null
    ? { event_type: "hf.dom.changed", payload }
    : { event_type: "hf.dom.changed", payload: { ...payload, request_id: requestId } };
}

// Strip value-bearing fields unless in development mode (§5.4). Pure: keys survive, values do not.
export function redact(payload, mode) {
  return mode === "development"
    ? payload
    : Object.fromEntries(Object.entries(payload).filter(([key]) => !VALUE_BEARING.includes(key)));
}

// Read X-Request-ID from a response header getter (§5.2). Pure: an absent or empty header is null.
export function readRequestId(getHeader) {
  return getHeader("X-Request-ID") || null;
}

// The one boundary: build the envelope from injected impure values and beacon it (§5.1). request_id is
// read fresh from the DOM through deps.requestId, so no module state holds it between requests.
export function emitBrowserEvent(event, deps) {
  const envelope = browserEvent(
    event.event_type,
    deps.timestamp(),
    deps.sessionId(),
    redact(event.payload, deps.mode()),
    deps.uuid(),
    deps.requestId(),
  );
  deps.sendBeacon(deps.endpoint ?? INGEST_ENDPOINT, JSON.stringify(envelope));
}
